plugins/events-backend/src/service/hub/DatabaseEventBusStore.ts

```diff
@@ -0,0 +1,555 @@
+      // events where read, the last ID out of all events
+      .update({
+        read_until: this.#db.raw(
+          'COALESCE(last_event_id, (SELECT MAX(id) FROM event_bus_events), 0)',
+        ),
+      })
+      .updateFrom({
+        events_array: 'events_array',
+        last_event_id: 'last_event_id',
+      })
+      .where(`${TABLE_SUBSCRIPTIONS}.id`, id)
+      .returning<[{ events: EventsRow[] }]>('events_array.events');
+
+    if (result.length === 0) {
+      throw new NotFoundError(`Subscription with ID '${id}' not found`);
+    } else if (result.length > 1) {
+      throw new Error(
+        `Failed to read subscription, unexpectedly updated ${result.length} rows`,
+      );
+    }
+
+    const rows = result[0].events;
+    if (!rows || rows.length === 0) {
+      return { events: [] };
+    }
+
+    return {
+      events: result[0].events.map(row => {
+        const { payload, metadata } = JSON.parse(row.data_json);
+        return {
+          topic: row.topic,
+          eventPayload: payload,
+          metadata,
+        };
+      }),
+    };
+  }
+
+  async setupListener(
+    subscriptionId: string,
+    options: {
+      signal: AbortSignal;
+    },
+  ): Promise<{ waitForUpdate(): Promise<{ topic: string }> }> {
+    const result = await this.#db<SubscriptionsRow>(TABLE_SUBSCRIPTIONS)
+      .select('topics')
+      .where({ id: subscriptionId })
+      .first();
+
+    if (!result) {
+      throw new NotFoundError(
+        `Subscription with ID '${subscriptionId}' not found`,
+      );
+    }
+
+    options.signal.throwIfAborted();
+
+    return this.#listener.setupListener(
+      new Set(result.topics ?? []),
+      options.signal,
+    );
+  }
+
+  #cleanup = async () => {
+    try {
+      const eventCount = await this.#db
+        .delete()
+        .from(TABLE_EVENTS)
+        // Delete any events that are outside both the min age and size window
+        .where('created_at', '<', new Date(Date.now() - this.#windowMinAge))
+        .andWhere(
+          'id',
+          '=',
+          this.#db
+            .raw(
+              this.#db
+                .select('id')
+                .from(TABLE_EVENTS)
+                .orderBy('id', 'desc')
+                .offset(this.#windowSize),
+            )
+            .wrap('ANY(ARRAY(', '))'),
+        )
+        // If events are outside the max age they will always be deleted
+        .orWhere('created_at', '<', new Date(Date.now() - this.#windowMaxAge));
+      this.#logger.info(
+        `Event cleanup resulted in ${eventCount} old events being deleted`,
+      );
+    } catch (error) {
+      this.#logger.error('Event cleanup failed', error);
+    }
+
+    try {
+      // Delete any subscribers that aren't keeping up with current events
+      const subscriberCount = await this.#db
+        .delete()
+        .from(TABLE_SUBSCRIPTIONS)
+        .where('read_until', '<', (q: Knex.QueryBuilder) =>
+          q.select(this.#db.raw('MIN(id)')).from(TABLE_EVENTS),
```

Comment:
iirc that didn't work for some reason, but I'll try it again

Reply:
I extracted both the min and max out to a separate query, no race to worry about since the min/max is the query that picks the point at which to start/stop consuming events.